src/sentry/coreapi.py

```diff
@@ -348,14 +348,29 @@ def ensure_does_not_have_ip(self, data):
         if 'sdk' in data:
             data['sdk'].pop('client_ip', None)
 
-    def insert_data_to_database(self, data, start_time=None, from_reprocessing=False):
+    def insert_data_to_database(self, data, start_time=None,
+                                from_reprocessing=False, attachments=None):
         if start_time is None:
             start_time = time()
+
         # we might be passed some sublcasses of dict that fail dumping
         if isinstance(data, DOWNGRADE_DATA_TYPES):
             data = dict(data.items())
+
+        cache_timeout = 3600
         cache_key = 'e:{1}:{0}'.format(data['project'], data['event_id'])
-        default_cache.set(cache_key, data, timeout=3600)
+        default_cache.set(cache_key, data, cache_timeout)
+
+        # Attachments will be empty or None if the "event-attachments" feature
+        # is turned off. For native crash reports it will still contain the
+        # crash dump (e.g. minidump) so we can load it during processing.
+        if attachments is not None:
+            for index, attachment in enumerate(attachments):
+                attachment_data = attachment.pop('data')
+                attachment_key = '%s:a:%s'.format(cache_key, index)
+                default_cache.set(attachment_key, attachment_data, cache_timeout, raw=True)
```

Comment:
Isn't this potentially like, 100MB of data? Or whatever limit we'll have for minidump ingestion?

Reply:
@mattrobenolt potentially but unlikely.

Reply:
Looking at this, I'd almost rather just directly write this into a `File` row during ingestion. I don't really think it'd be much worse to write directly into storage vs piping 100MB into cache, _then_ writing it into storage.

It's probably even faster to just write directly to disk given that we now use GCS. And would be less bookkeeping.

Reply:
@mattrobenolt the vast majority of minidumps will not be persisted. As such these files would only live for a very short period of time to be deleted immediately afterwards.

Reply:
Also more importantly: most minidumps are below 1MB.

Reply:
Where in the code here are they ever not being persisted? The only path I see that'd happen is if they don't ave the feature switch enabled, and we could just check that here instead.

Reply:
Right now all minidump processing happens in store. This is not going to work for CFI processing so we're temporarily pushing the minidump into the cache so we can pick it up in the processing task.

Independently of that we then *optionally* take that minidump and shove it into a `File` if a user has that feature enabled.

So the two cases we are about are that the vast majority of minidumps in the future will require CFI processing (all events?) so we need to move that out of the main store and into the processing (CFI processing requires debug symbols) but only a small percentage (~5% of minidump events?) will persist to a file.

Reply:
Yeah, I understand that even if they're not large. I'm just trying to understand why there's any benefit at all of going through cache before saving. The only reason I'd think that's valuable is if writing to cache would be faster. But given that it's writing data over a network into a single threaded process o the other side, and each of these are done sequentially in a loop, I don't think it's really that much more efficient vs just plowing directly into storage.

I could be wrong in assuming there's some other reason for this, but it seems like extra work to write into cache, then pull out later, and always attempt to clean up more keys.

Now, lastly, even if we do still think it's worthwhile to go through cache, our cache backend has a currently hardcoded limit of 50MB. So even in a condition where 100MB is rare, if anything over 50MB comes through, it'll hard be rejected. https://github.com/getsentry/sentry/blob/master/src/sentry/cache/redis.py#L23

Obviously this could be lifted, but I'd rather not since we've had incidents where this breaks Redis from super large values.

Reply:
If it's the case that we want/need to push these through Redis, I'd really advocate we use a different redis cluster then. Otherwise, we risk breaking event ingestion for all of Sentry if large blobs do get shoved through, no matter how rare they are, it'll happen, and it'll make everything else unstable.

Reply:
> I'm just trying to understand why there's any benefit at all of going through cache before saving. The only reason I'd think that's valuable is if writing to cache would be faster.

It's not just the writing that is significantly faster, but also it has TTLing and fast deletes.

> Now, lastly, even if we do still think it's worthwhile to go through cache, our cache backend has a currently hardcoded limit of 50MB. So even in a condition where 100MB is rare, if anything over 50MB comes through, it'll hard be rejected.

I would propose we just have a hard limit on minidumps enforced in sentry in addition to the load balancer for now.

> If it's the case that we want/need to push these through Redis, I'd really advocate we use a different redis cluster then.

I'm open to that for as long as we manage to have a reasonable fallback to the default sentry cache if not configured.

Reply:
Yeah, so this is fine then imo. I think we can enforce the size in Sentry code too. Doesn't change too much though.

I'm going to rope in @JTCunning then on how we want to explicitly go about this, but my assumption here is we don't want to shove these into current redis hosts.

And our preference would be to leverage redis directly, instead of going through `default_cache` (since this doesn't have the ability to go to another cluster).

And then we can use `redis-cluster`.

I'm not sure if there's an existing cluster we'd wanna use, but we can't shove this through our main pipeline.

> as long as we manage to have a reasonable fallback to the default sentry cache if not configured.

I assume you mean for on-premise, in which case, I think that's fine.